### ISSUE TITLE
Modify parse flags to include parsed date parts

### ISSUE THREAD

**Issue by maggiepint** [moment/moment#3037]
The whole Klingon date replace chrome extension leads me to an interesting issue.

I have a regex looking for dates that will match a whole bunch of formats:
Mar 1 2016
Mar 2016
1 Mar 2016
1 Mar
Mar 1
March 1 2016
March 2016
1 March 2016
1 March
March 1

By using the heck out of multiple-format parsing, I can mostly get these to parse.
I end up with an issue when I want to format them back out though. I don't have a good way of knowing which date parts were actually matched by the parser.
I can of course access internal variables like _f or _pf to get some of this info, but that's wrong to do and I won't do it :-).
It would be super convenient to be able to do something like ``.format('LL', true)``, with the bool being an indicator to purposely drop out unmatched date parts.

The result would be something like:
```js
moment('Mar 7 2012', 'MMMMDDY').format('MMMM DD YYYY', true)
"March 07 2012"

moment('Mar 7', 'MMMMDDY').format('MMMM DD YYYY', true)
"March 07"
//currently this would result in "March 07 2016"
```

I feel like there are actually a lot of reasonable use cases for this.

I'd be happy to play with making this work, but I'm looking for comments on whether people like the idea before I do it.

**Comment by maggiepint** [comment]
Follow up on this - I believe the same internal code required to make this work would enable us to easily address #2986 and #2987

**Comment by icambron** [comment]
FWIW, `_pf` is publicly exposed via `parsingFlags()`, though it doesn't quite have what you want. think it's 

On your proposal, I think it's too a rare use case to warrant that. I'd rather enhance `parsingFlags` to expose the right things, so you can do something like

```
var m = moment('Mar 7', ['MMMMDDY', ...]);
m.format(m.parsingFlags().formatUsed);
```

or similar.

**Comment by maggiepint** [comment]
I should get better at reading the docs.
Anyways, I'll agree that it's a fairly obscure use case. I actually have at least one situation where I could use it at work, but maybe we do strange things, I don't know.

To me, what is of importance here is what date parts were parsed - not necessarily the format used or the specific tokens used. We could simply save the version of _a that exists before we fill out missing parts with the current date. That could be saved on _pf and would generally solve my issue as I could write some simple ternaries to construct the format. It could also be used to mark dates invalid that have no useful parts parsed (#2986).

Maybe I'll go down that road.

**Comment by icambron** [comment]
Yeah, I definitely think that pf should be responsible for providing you any details of the mechanics of the parse, like which parts were defaulted, what format was used, and pretty much anything else you could need. So I'm +1 on adding features to it.

**Comment by maggiepint** [comment]
Any strong feelings on changing isValid to be false if there's nothing in this array, but there is a string value for _i? IMO - a date where we didn't parse anything but meridiem is totally invalid. It is a sort of breaking change though.

**Comment by mj1856** [comment]
I'm fine with @icambron's suggestion of exposing via paring flags, but not on changing the behavior of `format`.  I'm not really sure I understand the question about changing `isValid`.

**Comment by maggiepint** [comment]
Yeah, forget the behavior of format - the cool thing about bad ideas is that you don't have to implement them :-)

As far as the validity issue, it's just a mental step I'm taking based on this issue:
```js
moment('asdfgh', 'hh:mm a').isValid()
true
```
If I preserve the list of date parts that are actually defined by the parse, then this situation becomes solvable. Suppose I name that array _parsedParts (for the sake of argument). Remember that the internal value for the original input is _i.
Add a condition to isValid:
```javascript
!(_i && isNan(_i) && _parsedParts.length === 0)
```
Basically, the date is valid if the original input is defined, is a string, and we didn't parse any date parts.
This is imperfect code because JavaScript's understanding of types is weird, you can parse an X token with trailing letters after it, etc, but it feels like if I played with it enough, it would be solvable (perhaps famous last words there).

Maybe at this point the right thing to do is to just work on it instead of trying to explain it and see if it even can be done.

**Comment by maggiepint** [comment]
Actually, I'm making this way to hard with the input string. There is already an existing value for _meridiem that could be added to the parsing flags.

then in isValid:

```js
(!flags._meridiem || (flags._meridiem && flags.parsedParts.length>0))
```
The date is valid if the meridiem was not parsed, or if the meridiem was parsed and some other date part was parsed as well.

That solves the problem in a fairly simple manner. Only issue would be if someone used the format 'x a' or something - but we can't fix everything.

**Comment by maggiepint** [comment]
PR #3045
